fe/fe-core/src/main/java/com/starrocks/connector/paimon/PaimonMetadata.java

```diff
@@ -287,4 +296,89 @@ public long getTableCreateTime(String dbName, String tblName) throws Exception {
+        Long latestId = fileMonitorTable.snapshotManager().latestSnapshotId();
+        long latestSnapshotId = latestId == null ? Long.MIN_VALUE : latestId;
+        LOG.debug("Paimon table {} latest snapshotId {}, currentId {}",
+                paimonTable.getName(), latestSnapshotId, nextSnapshotId);
+        if (nextSnapshotId >= latestSnapshotId && latestSnapshotId != Long.MIN_VALUE) {
+            LOG.info("Paimon table {} currentId {} > latest snapshotId {} ",
+                    paimonTable.getName(), latestSnapshotId, nextSnapshotId);
+            return partitionToSnapshotId;
+        }
+        ReadBuilder readBuilder = fileMonitorTable.newReadBuilder();
+        StreamTableScan scan = readBuilder.newStreamScan();
+        TableRead read = readBuilder.newRead();
+        if (nextSnapshotId != Long.MIN_VALUE) {
+            scan.restore(nextSnapshotId + 1);
+        } else {
+            scan.restore(null);
+        }
+        try {
+            while (true) {
+                if (!scanMonitorTable(paimonTable, scan, read, partitionToSnapshotId)) {
+                    break;
+                }
+            }
+        } catch (OutOfRangeException e) {
+            partitionToSnapshotId.clear();
+            List<String> parts = listPartitionNames(paimonTable.getDbName(), paimonTable.getTableName());
+            for (String part : parts) {
+                partitionToSnapshotId.put(part, latestSnapshotId);
+            }
+            LOG.info("Paimon snapshot id {} has been out of date, return all latest partitions with latest id {}.",
+                    nextSnapshotId, latestSnapshotId
+            );
+        }
+        return partitionToSnapshotId;
+    }
+
+    private boolean scanMonitorTable(PaimonTable paimonTable, StreamTableScan scan, TableRead read,
+                             Map<String, Long> partitionToSnapshotId) {
+        TableScan.Plan plan = scan.plan();
+        if (plan.splits().isEmpty()) {
+            return false;
+        }
+        try {
+            read.createReader(plan).forEachRemaining(new Consumer<InternalRow>() {
+                @Override
+                public void accept(InternalRow row) {
+                    try {
+                        FileMonitorTable.FileChange fileChange = FileMonitorTable.toFileChange(row);
+                        RowDataConverter converter = new RowDataConverter(paimonTable.getNativeTable().
+                                schema().logicalPartitionType());
+                        List<String> partitionValues = converter.convert(fileChange.partition(),
+                                paimonTable.getPartitionColumnNames());
+                        String partition = FileUtils.makePartName(paimonTable.getPartitionColumnNames(), partitionValues);
+                        partitionToSnapshotId.put(partition, scan.checkpoint());
+                    } catch (IOException e) {
+                        LOG.error("Get fileChange failed.", e);
```

Comment:
this exception may be ignored?

Reply:
Query rewrite depends on this function to check if mv is out of date, so exception should be catch in case that normal query be killed by rewrite rule.
I will modify this code, throw this exception and catch by upper level function, then return all partition to avoid rewrite to stale mv.